ghproxy/ghcache/ghcache.go

```diff
@@ -232,37 +244,116 @@ func NewDiskCache(delegate http.RoundTripper, cacheDir string, cacheSizeGB, maxC
 			maxConcurrency,
 		)
 	}
+
+	go func() {
+		for range time.NewTicker(cachePruneInterval).C {
+			prune(cacheDir)
+		}
+	}()
 	return NewFromCache(delegate,
-		func(partitionKey string) httpcache.Cache {
+		func(partitionKey string, expiresAt time.Time) httpcache.Cache {
+			basePath := path.Join(cacheDir, "data", partitionKey)
+			tempDir := path.Join(cacheDir, "temp", partitionKey)
+			if err := writecachePartitionMetadata(basePath, tempDir, expiresAt); err != nil {
+				logrus.WithError(err).Warn("Failed to write cache metadata file, pruning will not work")
+			}
 			return diskcache.NewWithDiskv(
 				diskv.New(diskv.Options{
-					BasePath:     path.Join(cacheDir, "data", partitionKey),
-					TempDir:      path.Join(cacheDir, "temp", partitionKey),
+					BasePath:     basePath,
+					TempDir:      tempDir,
 					CacheSizeMax: uint64(cacheSizeGB) * uint64(1000000000), // convert G to B
 				}))
 		},
 		maxConcurrency,
 	)
 }
 
+func prune(baseDir string) {
+	// All of this would be easier if the structure was base/partition/{data,temp}
+	// but because of compatibility we can not change it.
+	for _, dir := range []string{"data", "temp"} {
+		base := path.Join(baseDir, dir)
+		cachePartitionCandidates, err := os.ReadDir(base)
+		if err != nil {
+			logrus.WithError(err).Warn("os.ReadDir failed")
+			// no continue, os.ReadDir returns partial results if it encounters an error
+		}
+		for _, cachePartitionCandidate := range cachePartitionCandidates {
+			if !cachePartitionCandidate.IsDir() {
+				continue
+			}
+			metadataPath := path.Join(base, cachePartitionCandidate.Name(), cachePartitionMetadataFileName)
+
+			// Read optimistically and return errors
+			raw, err := ioutil.ReadFile(metadataPath)
+			if err != nil {
+				continue
+			}
+			var metadata cachePartitionMetadata
+			if err := json.Unmarshal(raw, &metadata); err != nil {
+				logrus.WithError(err).WithField("filepath", metadataPath).Error("failed to deserialize metadata file")
+				continue
+			}
+			if metadata.ExpiresAt.After(time.Now()) {
+				continue
+			}
+			paritionPath := filepath.Dir(metadataPath)
+			logrus.WithField("path", paritionPath).WithField("expiresAt", metadata.ExpiresAt.String()).Info("Cleaning up expired cache parition")
+			if err := os.RemoveAll(paritionPath); err != nil {
+				logrus.WithError(err).WithField("path", paritionPath).Error("failed to delete expired cache parition")
+			}
+		}
+	}
+}
+
+func writecachePartitionMetadata(basePath, tempDir string, expiresAt time.Time) error {
+	if expiresAt.IsZero() {
+		return nil
```

Comment:
Won't this lead to leaks? Why not failsafe to writing metadata that expires at `time.Now().Add(time.Hour)` or something?

Reply:
No, it won't and we can't do that. The whole reason the expiry information is passed on from the client is tokens validity varies and in the case of PAT, it never expires which will gets by an empty `expiresAt`. If we unconditionally added 1h here, we would always delete all caches after one hour, we cant do that.

Reply:
Ah, I see - this is to handle PAT. Perhaps a comment would be good to clarify this since it's implicit behavior?

Reply:
I guess I would have expected "no expiry header" -> "no call to writing metadata" rather than "no expiry header" -> "pass an invalid date" -> "do nothing" but maybe that's just me being confused by it all

Reply:
Also I guess it would not hurt to have a default cache TTL for PAT entries, too, since they could hit the same issues that apps auth hits, on a smaller machine or with fewer inodes free? Setting the TTL to a week or something should not cause adverse effects.

Reply:
Added a comment and made it a pointer to further clarify this might not be set. The TTL is for the entire cache, not individual entries so we can never evict a PAT cache